documentation/development-docs/design-guidelines/cmdlet-best-practices.md

```diff
@@ -0,0 +1,160 @@
+
+The code below showshow this should look in a cmdlet:
+
+```cs
+[Cmdlet(..., SupportsShouldProcess = true), OutputType(typeof(...))]
+public class MySampleCmdlet : MyBaseCmdlet
+{
+    // parameters omitted
+
+    public override void ExecuteCmdlet()
+    {
+        // other code omitted
+
+        if (ShouldProcess(targetResource, actionMessage))
+        {
+            // make the change
+        }
+    }
+}
+```
+
+More information about `ShouldProcess` can be found in the [_Should Process and Confirm Impact_](./should-process-confirm-impact.md) document.
+
+#### When to Add the Force Parameter
+
+The `-Force` parameter is reserved for special scenarios where additional confirmation from the user is required. From the above document on [_Should Process and Confirm Impact_](./should-process-confirm-impact.md) document:
+
+> _Some cmdlets require additional confirmation. For example, if a cmdlet would destroy existing resources in some circumstances, the cmdlet might detect that condition and prompt the user to verify before continuing. Overwriting an existing resource during resource creation, overwriting a file when downloading data, deleting a resource that is currently in use, or deleting a container that contains additional resources are all example of this pattern. To implement additional confirmation, and allow scripts to opt out of additional prompts, the above pattern is enhanced with calls to `ShouldContinue()` and the `-Force` parameter._
+
+### `AsJob`
+
+All long running operations must implement the `-AsJob` parameter, which will allow the user to create jobs in the background. For more information about PowerShell jobs and the `-AsJob` parameter, read [this doc](https://docs.microsoft.com/en-us/powershell/azure/using-psjobs).
+
+To implement the `-AsJob` parameter, simply add the parameter to the end of the parameter list:
+
+````cs
+[Parameter(Mandatory = false, HelpMessage = "Run cmdlet in the background")]
+public SwitchParameter AsJob { get; set; }
+````
+
+Once you add the parameter, please manually test that the job is created and successfully completes when the parameter is specified.  Additionally, please ensure that the help files are updated with this parameter.
+
+To ensure that `-AsJob` is not broken in future changes, please add a test for this parameter. To update tests to include this parameter, use the following pattern:
+
+````powershell
+$job = Get-AzSubscription -AsJob
+$job | Wait-Job
+$subcriptions = $job | Receive-Job
+````
+
+To set a custom job name, please use [`SetBackgroupJobDescription`](https://github.com/Azure/azure-powershell-common/blob/master/src/Common/AzurePSCmdlet.cs#L810). The default job description is: "Long Running Operation for '{cmdlet name}' on resource '{resource name}'"
+
+### Required Parameter Sets
+
+In most Azure PowerShell cmdlets, there is a bare minimum of three parameter sets that need to be implemented.
```

Comment:
Can you add a link to the piping in powershell doc?

Reply:
@maddieclayton linked out to the piping document in both corresponding scenarios below